Extend TestingGuide (adds notes on formatting) (#216)
llvm/mlir-www: website/content/getting_started/TestingGuide.md

}
```

### Test Formatting Best Practices

When adding new tests, strive to follow these two key rules:

> 1. **Follow the existing naming and whitespace style.**

Reviewer: When you say 'existing style', what scope do you have in mind?

Author: I’m deliberately avoiding specifying this :)

Making things globally consistent would be great for... consistency. However, that assumes **one size fits all**, which isn't always realistic. Given the current state of things, I also feel this would be an impractical goal.
The same applies to **per-file** vs **per-directory** consistency. For example, "mlir/test/Dialect/Vector" has around 50 files - enforcing 100% consistency across them would require significant effort and churn.

My view is that contributors and reviewers should make the right call on a case-by-case basis while being mindful of:

* Reinventing the wheel in terms of formatting.
* New tests ignoring existing conventions (whether at the file or directory level).

Ultimately, I see this as a **gradual** process. Let’s start with these guidelines to help "correct the course." Over time, we can refine them - possibly transitioning from **guideline** to requirement if it makes sense.

WDYT?

Reviewer: I've worked on a few different codebases that approached this differently: some preferred local consistency, some global (e.g., google3). I agree that this doesn't have to one or the other and there's some room for interpretation, but at the same time I'd like to have some codified incentive that will push us towards **higher** quality tests/code.

For example, we could say we prefer global consistency unless the local standards are either higher or more specialized. If a PR adheres to the global standards but sticks like a sore thumb compared to other local code, I'd think that in most cases sending a separate clean up PR would be a reasonable request.

Author: Thanks, Jakub! Sorry for the delayed reply—I haven’t had much time for this PR today. 😅  

My overall sentiment is that we should approach defining these guidelines as an **incremental process**, avoiding overly strict or prescriptive rules.  

> For example, we could say we prefer global consistency unless the local standards are either higher or more specialized.

I personally like that, but I worry that others may disagree, and we might get stuck discussing it. Also, if we ask for **global consistency**, who is actually going to help us get there with what we have today?  

We shouldn't require new contributions to be **fully consistent** with everything we already have, given how... inconsistent our tests are. 😅  

---

### **Proposed Levels of Consistency**
How about defining **different levels of consistency**?  

1. **Per-PR consistency**  
   - When adding new Ops, documentation, and tests, they should use an identical style.  
   - **Mandatory**.  

2. **Per-file consistency**  
   - If a file is already consistent, maintain that.  
   - If a file is inconsistent, create a GitHub issue to address it.  
   - **Mandatory**.  

3. **Per-directory consistency**  
   - This might require a lot of churn, so I’m unsure about enforcing it strictly.  
   - **Nice-to-have**.  

4. **Per-project (MLIR-wide) consistency**  
   - Achieving this would require **volunteers** to refactor existing tests.  
   - **Nice-to-have**.  

---

This is just a rough idea - suggestions are welcome!

Reviewer: > My overall sentiment is that we should approach defining these guidelines as an incremental process, avoiding overly strict or prescriptive rules.
> I personally like that, but I worry that others may disagree, and we might get stuck discussing it. 

This is a very good point, let's revisit this once the initial version lands.

> Also, if we ask for global consistency, who is actually going to help us get there with what we have today?
> We shouldn't require new contributions to be fully consistent with everything we already have, given how... inconsistent our tests are.

Maybe 'consistency' is not the best term here. The idea would be that we would be getting closer to the agreed on 'testing guidelines' (similar to the [coding standards](https://llvm.org/docs/CodingStandards.html)) incrementally as we land new code, and refactor old tests along the way. Having the coding standards codified and linkable is a fantastic tool in code reviews because it allows you to point out the best practices without making it feel like you are making one-of arguments against someone's implementation. I'd hope that with testing guidelines we can accomplish something similar and gradually steer tests towards higher maintainability/quality/readability.

Reviewer: IE, my proposed rule of thumb would be something along the lines: 'Follow the testing guidelines while adhering to the local naming and formatting style.'

> 1. **Follow the existing naming and whitespace style.**

Reviewer 2: nit: Should the existing bad style (test_1, test_2, ...) prevent naming new test cases more reasonably?

//  CHECK-SAME:       %[[A1:.*]]: vector<16xi8>) -> vector<16xi8> {
//       CHECK:   %[[C0:.*]] = arith.constant 0 : index
//       CHECK:   %[[LOAD:.*]] = vector.load %[[A0]][%[[C]]] : memref<16xi8>, vector<16xi8>
//       CHECK:   return %[[LOAD]] : vector<16xi8>
func.func @maskedload_regression_2(%arg0: memref<16xi8>, %arg1: vector<16xi8>) -> vector<16xi8> {
  %c0 = arith.constant 0 : index
  %vec_i1 = vector.constant_vec_i1 [16] : vector<16xi1>

  %ld = vector.maskedload %arg0[%c0], %vec_i1, %arg1
    : memref<16xi8>, vector<16xi1>, vector<16xi8> into vector<16xi8>

  return %ld : vector<16xi8>
}

// CHECK-LABEL:   func @maskedload_regression_3(
// CHECK-SAME:        %[[A0:.*]]: memref<16xf32>,
// CHECK-SAME:        %[[A1:.*]]: vector<16xf32>) -> vector<16xf32> {
//      CHECK:    return %[[A1]] : vector<16xf32>
func.func @maskedload_regression_3(%arg0: memref<16xf32>, %arg1: vector<16xf32>) -> vector<16xf32> {
  %c0 = arith.constant 0 : index
  %vec_i1 = vector.constant_vec_i1 [0] : vector<16xi1>

  %ld = vector.maskedload %arg0[%c0], %vec_i1, %arg1
    : memref<16xf32>, vector<16xi1>, vector<16xf32> into vector<16xf32>

  return %ld : vector<16xf32>

> }

Reviewer: Do we need all 3 examples to drive your point or could we limit this to one or two?

Author: "Generalization Takes Three Examples" :)

It allows me to go into a bit of nuance in this paragraph (i.e. to identify the commonalities):

```mlir
  ##### Step 2: Improve Test Naming

  Instead of using "regression" (which doesn't add unique information), rename
  tests based on key attributes:

  * All examples test the `vector.maskedload` to `vector.load` lowering.
  * The first test uses a _dynamically_ shaped `memref`, while the others use
    _static_ shapes.
  * The mask in the first two examples is "all true" (`vector.constant_mas
    [16]`), while it is "all false" (`vector.constant_mask [0]`) in the thir
    example.
  * The first and the third tests use `i32` elements, whereas the second uses
    `i8`.
```

That's not something I feel super strongly about. Usually its tricky to see the "optimal" solution on first iteration and I guess that's the case here.

> ```

Reviewer 3: Nit: newline after block. I think our renderer is particular about it.

  %mask = vector.constant_mask [16] : vector<16xi1>

  %ld = vector.maskedload %base[%c0], %mask, %pass_thru
    : memref<16xi8>, vector<16xi1>, vector<16xi8> into vector<16xi8>

  return %ld : vector<16xi8>
}

// CHECK-LABEL:   func @maskedload_regression_3(
// CHECK-SAME:        %[[BASE:.*]]: memref<16xf32>,
// CHECK-SAME:        %[[PASS_THRU:.*]]: vector<16xf32>) -> vector<16xf32> {
//      CHECK:    return %[[PASS_THRU]] : vector<16xf32>
func.func @maskedload_regression_3(%base: memref<16xf32>, %pass_thru: vector<16xf32>) -> vector<16xf32> {
  %c0 = arith.constant 0 : index
  %mask = vector.constant_mask [0] : vector<16xi1>

  %ld = vector.maskedload %base[%c0], %mask, %pass_thru
    : memref<16xf32>, vector<16xi1>, vector<16xf32> into vector<16xf32>

  return %ld : vector<16xf32>
}
```

##### Step 2: Improve Test Naming

Instead of using "regression" (which doesn't add unique information), rename
tests based on key attributes:

* All examples test the `vector.maskedload` to `vector.load` lowering.
* The first test uses a _dynamically_ shaped `memref`, while the others use
  _static_ shapes.
* The mask in the first two examples is "all true" (`vector.constant_mas

>   [16]`), while it is "all false" (`vector.constant_mask [0]`) in the thir

Reviewer 4: ```suggestion
  [16]`), while it is "all false" (`vector.constant_mask [0]`) in the the
```

  example.

> * The first and the third tests use `i32` elements, whereas the second uses

Reviewer 2: The key attributes should reflect what's important for the tested transform, in this case having different data types is good for completeness but they don't trigger any special/separate logic.
It would be sufficient to highlight one case that uses different data type with, for example, a suffix and omit data type for others.

I know that's not the intention behind this example but it can be misused to push for "enterprisification" (classic java gem: https://projects.haykranen.nl/java/) of naming with combinatorial explosion of all the little details.

Author: Added a short paragraph in this [commit](https://github.com/llvm/mlir-www/pull/216/commits/6f7b7697ac7911f935fa9db047a4f6e594b6fe06). Let me know what you think - do we need more?

Reviewer 2: Great clause 👍
It conveys the spirit of these guidelines and defends against too pedantic rule lawyers.

The update in Step 2 made us realize that we were missing an important edge
case:

* A mask that is neither "all true" nor "all false".

Unlike the existing cases, this mask must be preserved. In this scenario,
`vector.load` is not the right abstraction. Thus, no lowering should occur:

```mlir
// CHECK-LABEL:   func @negative_maskedload_to_load_static_i32_mixed(
// CHECK-SAME:        %[[BASE:.*]]: memref<16xf32>,
// CHECK-SAME:        %[[PASS_THRU:.*]]: vector<16xf32>) -> vector<16xf32> {
//      CHECK:    vector.maskedload
func.func @negative_maskedload_to_load_static_i32_mixed(%base: memref<16xf32>, %pass_thru: vector<16xf32>) -> vector<16xf32> {
  %c0 = arith.constant 0 : index
  %mask = vector.constant_mask [4] : vector<16xi1>

  %ld = vector.maskedload %base[%c0], %mask, %pass_thru
    : memref<16xf32>, vector<16xi1>, vector<16xf32> into vector<16xf32>

  return %ld : vector<16xf32>
}
```

The `negative_` prefix indicates that this test should fail to lower, as the
pattern should not match.

To summarize, here’s the naming convention used:

* `@{negative_}?maskedload_to_load_{static|dynamic}_{i32|i8}_{all_true|all_false|mixed}`.

> #### Final Points - key principles

Reviewer: ```suggestion
#### Final Points - Key Principles
```

The above approach is just an example. It may not fit your use case perfectly,
so feel free to adapt it as needed. For example, for "folding" tests, it may
make more sense to use "no" as prefix for negative tests (e.g.
`@no_fold_<case>_<subcase>`). Key principles to follow:

* Make tests self-documenting.
* Follow existing conventions.

By applying these best practices, we leverage available tools (e.g., test

> function names) to make tests easier to discover and maintain.

Reviewer 5: This a good addition: I would still add some documentation in comment to a test: it's not always clear just from a test name what's going on, and the same logic as in code applies (document the why, the context, the "what's no obvious from the code").

Author: Yes, comments are also super important, thanks for raising this!

I have some ideas on "comments", so let me treat this as encouragement to extend this PR with a dedicated section on documentation :)

//       CHECK:   %[[LOAD:.*]] = vector.load %[[BASE]][%[[C]]] : memref<16xi8>, vector<16xi8>
//       CHECK:   return %[[LOAD]] : vector<16xi8>
func.func @maskedload_to_load_static_i8_all_true(%base: memref<16xi8>, %pass_thru: vector<16xi8>) -> vector<16xi8> {
  %c0 = arith.constant 0 : index
  %mask = vector.constant_mask [16] : vector<16xi1>

  %ld = vector.maskedload %base[%c0], %mask, %pass_thru
    : memref<16xi8>, vector<16xi1>, vector<16xi8> into vector<16xi8>

  return %ld : vector<16xi8>
}

// CHECK-LABEL:   func @maskedload_to_load_static_i32_all_false(
// CHECK-SAME:        %[[BASE:.*]]: memref<16xf32>,
// CHECK-SAME:        %[[PASS_THRU:.*]]: vector<16xf32>) -> vector<16xf32> {
//      CHECK:    return %[[PASS_THRU]] : vector<16xf32>
func.func @maskedload_to_load_static_i32_all_false(%base: memref<16xf32>, %pass_thru: vector<16xf32>) -> vector<16xf32> {
  %c0 = arith.constant 0 : index
  %mask = vector.constant_mask [0] : vector<16xi1>

  %ld = vector.maskedload %base[%c0], %mask, %pass_thru
    : memref<16xf32>, vector<16xi1>, vector<16xf32> into vector<16xf32>

  return %ld : vector<16xf32>
}
```

##### Step 3: Add The Newly Identified Missing Case

> The update in Step 2 made us realize that we were missing an important edge

Reviewer 6: I don't think the we that includes me the reader realised anything here to be honest, you're about to tell me what you realised, the author of this.

Rephrase to avoid this.

"Step 2 made it possible to see that there is a case which is not tested."

* **Write Orthogonal Tests**
Struggling with good naming may indicate the absence of a clear pattern in the
set of tests being written. A good rule of thumb is to avoid testing the same
thing repeatedly. Before writing tests, define clear categories to cover (e.g.,
number of loops, data types). This often leads to a natural naming scheme—for
example: `@loop_depth_2_i32`.

* **What vs Why**
Test names should reflect _what_ is being tested, not _why_.

Encoding _why_ in test names can lead to unnecessarily long and complex names.
Instead, add inline comments where needed.

#### Don't forget the common sense

As a reviewer or contributor, always apply common sense when naming functions
and variables. Encoding too much information in names can negatively impact
readability and maintainability.

Trust your judgment. When in doubt, consult your future self: "Will this still
make sense to me six months from now?"

#### Final Points - Key Principles

The above approach is just an example. It may not fit your use case perfectly,
so feel free to adapt it as needed. For example, for "folding" tests, it may
make more sense to use "no" as prefix for negative tests (e.g.

> `@no_fold_<case>_<subcase>`). Key principles to follow:

Author: Good point, let me remove it from here. I do want to preserve `@no_fold_<case>_<subcase>` as an example though, so I will _move_ it elsewhere (as opposed to _removing_ it altogether). 

For context, @adam-smnk and I discussed that very example in https://github.com/llvm/llvm-project/pull/122449 (in fact, that prompted me to create this PR). In general, I think that it's good to have additional examples to demonstrate various options. Mostly to avoid people think that there's only one correct way of naming things. But we obviously also want to avoid going too far.

We can improve documentation further by clarifying what pattern is being
tested, providing high-level reasoning, and consolidating shared comments. For
instance:

```mlir
///----------------------------------------------------------------------------------------
/// [Pattern: TransferWritePermutationLowering]
///
/// IN: vector.transfer_write (_transposed_ minor identity permutation map)
/// OUT: vector.transpose + vector.transfer_write (minor identity permutation map)
///
/// Note, `permutation_map` from the input Op is replaced with the newly
/// inserted vector.traspose Op.
///----------------------------------------------------------------------------------------

// CHECK-LABEL:   func.func @xfer_write_minor_identity_transposed
//  CHECK-SAME:      %[[VEC:.*]]: vector<4x8xi16>,
//  CHECK-SAME:      %[[MEM:.*]]: memref<2x2x8x4xi16>
//  CHECK-SAME:      %[[IDX:.*]]: index)
//       CHECK:      %[[TR:.*]] = vector.transpose %[[VEC]], [1, 0] : vector<4x8xi16> to vector<8x4xi16>
// (...)
```

This documents:
* The transformation pattern being tested.
* The key logic behind the transformation.
* The expected change in output.

#### Documenting the "What"

> While we generally document why, documenting what is valid and encouraged in

Reviewer 6: Also if you use "you" it's more direct like:
"You should always document why, but also..."

//  CHECK-DAG:  %[[V_OUTPUT_R:.+]] = vector.transfer_read %[[OUTPUT]][%[[C0]], %[[C0]], %[[C0]]], %[[F0]]

//      CHECK:   %[[V_INPUT_0:.+]] = vector.extract_strided_slice %[[V_INPUT_R]]
//      CHECK:   %[[V_INPUT_1:.+]] = vector.extract_strided_slice %[[V_INPUT_R]]

//      CHECK:    %[[V_FILTER:.+]] = vector.extract %[[V_FILTER_R]][0] : vector<3x8xf32> from vector<1x3x8xf32>

//      CHECK:  %[[V_OUTPUT_0:.+]] = vector.extract_strided_slice %[[V_OUTPUT_R]]
//      CHECK:  %[[V_OUTPUT_1:.+]] = vector.extract_strided_slice %[[V_OUTPUT_R]]

/// w == 0, kw == 0
//      CHECK:   %[[CONTRACT_0:.+]] = vector.contract
// CHECK-SAME:     %[[V_INPUT_0]], %[[V_FILTER]], %[[V_OUTPUT_0]]

/// w == 1, kw == 0
//      CHECK:   %[[CONTRACT_1:.+]] = vector.contract
// CHECK-SAME:     %[[V_INPUT_1]], %[[V_FILTER]], %[[V_OUTPUT_1]]

/// w == 0, kw == 0
//      CHECK:   %[[RES_0:.+]] = vector.insert_strided_slice %[[CONTRACT_0]], %[[V_OUTPUT_R]]
/// w == 1, kw == 0
//      CHECK:   %[[RES_1:.+]] = vector.insert_strided_slice %[[CONTRACT_1]], %[[RES_0]]

/// Write the result back in one shot.
//      CHECK:   vector.transfer_write %[[RES_1]], %[[OUTPUT]][%[[C0]], %[[C0]], %[[C0]]]
```

> Note that while the comments document what is happening (e.g., "Write the

Reviewer 6: "Though the comments document..."

Is a neater way to start.

result back in one shot"), some variables — like `w` and `kw` — remain

> "enigmatic" and are not explicitly explained. This might leave us

Reviewer 6: "some variables — like `w` and `kw` — are not explained."

Simplify this. I know what you mean by explicitly, but you're about to talk about that in the next sentence anyway so no need to add more big words here :)

> second-guessing their meaning at first. However, that’s intentional - their

Reviewer 6: Cut the whole "This might leave us..." sentence. It just delays getting to the key point here from "However, that's...."

Also, we can skip the however.

"are not explained. This is intentional"

This makes the point clear and does not send the reader around the houses to get to it.

purpose becomes clear when studying the corresponding Linalg vectorizer
implementation (or, when analysing how `linalg.conv_1d_nwc_wcf` works).

> It’s important to remember that comments should assist in understanding the

Reviewer 6: "Comments help you understand code, they do not replace the need to read it. Comments guide the reader, they do not repeat what the code already says."

Drop the important, mostly importantly.

If it wasn't important you wouldn't bother to write it down, plus the average new reader of this has no relative scale of importance so everything written is important anyway (which is good, it keeps people diligent).

code, not replace the need to read it. Their role is to guide the reader, not
to restate what the code already conveys.

#### Final Takeaways

> * Prioritize documenting _why_ something happens, unless the _what_ is non-trivial.

Reviewer 6: "Always document why, document what if you need to" ?

> * Use high-level block comments to describe patterns being tested.

Reviewer 6: "Use block comments to describe the patterns being tested"

Otherwise I'm wondering what a high-level block comment is, is this different from a low-level block, where are my basement level comments at?

I think I know what you mean, but if you can be more straightforward here, again, it's one less chance for confusion.

> * Think about maintainability - comments should help future developers

Reviewer 6: I would add "(which includes you)" here because we can't stress that enough.

Stressing it won't teach the lesson - only experience really does that, but it helps.

  understand tests at a glance.

> * Comments should assist, not replace reading the code. Avoid over-explaining.

Reviewer 6: The last part seems like a bad transition, not quite a non-sequitur though. Make that into its own point?

* Avoid redundant explanations.

?

Reviewer 6: This does feel a bit like:
* do a lot of this
* but don't do too much
* got it?

But this is inherent to the process, and will come down to reviewer preference so I think this is just a feature of communication in general.

Author: It's very hard to find the right balance, yes. In fact, one of my goals is to hint that:

*  _Finding the right balance w.r.t. comments might be your most challenging task._

Let me rephrase this a bit to improve the flow. Perhaps I should re-write it altogether, but I need a bit more time to think about it. Alternatively, wait and see how people use these guidelines in practice 🤷🏻

  %c0 = arith.constant 0 : index
  %mask = vector.constant_mask [16] : vector<16xi1>

  %ld = vector.maskedload %base[%c0], %mask, %pass_thru
    : memref<16xi8>, vector<16xi1>, vector<16xi8> into vector<16xi8>

  return %ld : vector<16xi8>
}

// CHECK-LABEL:   func @maskedload_regression_3(
// CHECK-SAME:        %[[BASE:.*]]: memref<16xf32>,
// CHECK-SAME:        %[[PASS_THRU:.*]]: vector<16xf32>) -> vector<16xf32> {
//      CHECK:    return %[[PASS_THRU]] : vector<16xf32>
func.func @maskedload_regression_3(%base: memref<16xf32>, %pass_thru: vector<16xf32>) -> vector<16xf32> {
  %c0 = arith.constant 0 : index
  %mask = vector.constant_mask [0] : vector<16xi1>

  %ld = vector.maskedload %base[%c0], %mask, %pass_thru
    : memref<16xf32>, vector<16xi1>, vector<16xf32> into vector<16xf32>

  return %ld : vector<16xf32>
}
```

##### Step 2: Improve Test Naming

Instead of using "regression" (which doesn't add unique information), rename
tests based on key attributes:

* All examples test the `vector.maskedload` to `vector.load` lowering.
* The first test uses a _dynamically_ shaped `memref`, while the others use
  _static_ shapes.

> * The mask in the first two examples is "all true" (`vector.constant_mas

Reviewer 7: `mas` -> `mask`

/// w == 1, kw == 0
//      CHECK:   %[[RES_1:.+]] = vector.insert_strided_slice %[[CONTRACT_1]], %[[RES_0]]
// CHECK-SAME:     {offsets = [0, 1, 0], strides = [1, 1, 1]} : vector<4x1x8xf32> into vector<4x2x8xf32>

/// Write the result back in one shot.
//      CHECK:   vector.transfer_write %[[RES_1]], %[[OUTPUT]][%[[C0]], %[[C0]], %[[C0]]]

```

Note that while the comments document what is happening (e.g., "Write the
result back in one shot"), some variables — like `w` and `kw` — remain
"enigmatic" and are not explicitly explained. This might leave us
second-guessing their meaning at first. However, that’s intentional—their
purpose becomes clear when studying the corresponding Linalg vectorizer
implementation.

It’s important to remember that comments should assist in understanding the
code, not replace the need to read it. Their role is to guide the reader, not
to restate what the code already conveys.

#### Final Takeaways
* Prioritize documenting _why_ something happens, unless the _what_ is non-trivial.
* Use high-level block comments to describe patterns being tested.
* Think about maintainability - comments should help future developers

> 	understand tests at a glance.

Reviewer 7: weird indentation?

}
```

### Test Formatting Best Practices

When adding new tests, strive to follow these two key rules:

1. **Follow the existing naming and whitespace style.**

>    - This applies when modifying existing test files.

Author: I like your suggestion, thanks! Shall we also ask people to create GitHub issues when a test file _does not_ follow a consistent style?

CC @kuhar (I mentioned something similar in another thread with you)

2. **Consistently document the edge case being tested.**
   - Clearly state what makes this test unique and how it complements other
     similar tests.

> 

Reviewer 8: 4. **Names and Comments**
 Test names should reflect **what** is being tested. The **why** shouldnt be attempted to code in the name and unless obvious from code/context, must be added as comment but avoiding over-explainng..

> 

Author: These are great, thanks @javedabsar1 !  I'd like to keep this particular section very short (and limit the guidelines to two very high level points). Is it OK if I add these points in `#### What if there is no pre-existing style to follow?` instead?

> 

Reviewer 8: > These are great, thanks @javedabsar1 ! I'd like to keep this particular section very short (and limit the guidelines to two very high level points). Is it OK if I add these points in `#### What if there is no pre-existing style to follow?` instead?

Sounds good.

// CHECK-SAME:        %[[PASS_THRU:.*]]: vector<16xf32>) -> vector<16xf32> {
//      CHECK:    vector.maskedload
func.func @negative_maskedload_to_load_static_i32_mixed(%base: memref<16xf32>, %pass_thru: vector<16xf32>) -> vector<16xf32> {
  %c0 = arith.constant 0 : index
  %mask = vector.constant_mask [4] : vector<16xi1>

  %ld = vector.maskedload %base[%c0], %mask, %pass_thru
    : memref<16xf32>, vector<16xi1>, vector<16xf32> into vector<16xf32>

  return %ld : vector<16xf32>
}
```

The `negative_` prefix indicates that this test should fail to lower, as the
pattern should not match.

To summarize, here’s the naming convention used:

* `@{negative_}?maskedload_to_load_{static|dynamic}_{i32|i8}_{all_true|all_false|mixed}`.

#### What if there is no pre-existing style to follow?

If you are adding a new test file, you can use other test files in the same
directory as inspiration.

If the test file you are modifying lacks a clear style and instead has mixed,
inconsistent styles, try to identify the dominant one and follow it. Even
better, consider refactoring the file to adopt a single, consistent style —
this helps improve our overall testing quality.

> This is also encouraged when the existing style leaves room for improvement.

Reviewer 2: nit: Specify `this` more explicitly. Does it refer to following the existing style or refactoring?

so feel free to adapt it as needed. For example, for "folding" tests, it may
make more sense to use "no" as prefix for negative tests (e.g.
`@no_fold_<case>_<subcase>`). Key principles to follow:

* Make tests self-documenting.
* Follow existing conventions.

By applying these best practices, we leverage available tools (e.g., test
function names) to make tests easier to discover and maintain.

### Test Docmentation Best Practices

Apart from following good naming and formatting conventions, please document
your tests with comments. Focus on explaining **why** rather than **what**,
since the latter is usually clear from the code itself.

As an example, consider two test cases exercising the same pattern:

```mlir
// CHECK-LABEL:   func.func @xfer_write_minor_identity_transposed
//  CHECK-SAME:      %[[VEC:.*]]: vector<4x8xi16>,
//  CHECK-SAME:      %[[MEM:.*]]: memref<2x2x8x4xi16>
//  CHECK-SAME:      %[[IDX:.*]]: index)
//       CHECK:      %[[TR:.*]] = vector.transpose %[[VEC]], [1, 0] : vector<4x8xi16> to vector<8x4xi16>
//       CHECK:      vector.transfer_write

/// The permutation map was replaced with vector.transpose

> // CHECK-NOT:       permutation_map

Reviewer 2: This example is also made by the second example. Could they be collapsed?

Also, while the previous 3 examples at least are driven by more detailed narrative, I feel these verbose examples are overkill for a message that is summarized with two bullet points (just to be clear - I mean smaller example is more readable, not adding more description ;) ).

Or perhaps the key parts could use some emphasis (bold etc.). Or the actual test details could be omitted for documentation clarity.
Blob of IR triggers tl;dr before I filter out the message.

Reviewer 2: Same for the convolution example.

Author: Agreed, let me trim these examples. Let me know what you think.

> Or perhaps the key parts could use some emphasis (bold etc.).

Is it possible in code blocks in Markdown?

> Same for the convolution example.

Yes, it's quite long. It's a bit tricky to find good examples that are also short 🤔 But I totally agree that examples in this guide should be of highest possible quality. No pressure! 😅

Reviewer 2: > > Or perhaps the key parts could use some emphasis (bold etc.).
> 
> Is it possible in code blocks in Markdown?

Not certain but most likely no (in standard Markdown).
Docs and examples like these are always art of tradeoffs. Do these the examples need to be complete and self-contained? Or are they primarily a vehicle to guide (reader's gaze) through key concepts?

I like the cleanups for `Test Documentation Best Practices` section. Each code block has is a self contained example that showcases a concept with matching description.

In case of `Example: Improving Test Readability & Naming`, I feel it's hard to follow changes done with each step. Especially as each code block fills up my screen, I have to jump between current block, previous block, and the step description.

Since formatting is not really an option, maybe emphasis by omission of the common elements could help? Sth like:
Base case:
```mlir
// CHECK-LABEL:   func @maskedload_regression_1(
//  CHECK-SAME:       %[[A0:.*]]: memref<?xf32>,
//  CHECK-SAME:       %[[A1:.*]]: vector<16xf32>
func.func @maskedload_regression_1(%arg0: memref<?xf32>, %arg1: vector<16xf32>) {
  ...
  //       CHECK:   %[[LOAD:.*]] = vector.load %[[A0]][%[[C]]] : memref<?xf32>, vector<16xf32>
  %ld = vector.maskedload %arg0[%c0], %vec_i1, %arg1
  ...
}
```
Step 1 - Use Consistent Variable Names:
```mlir
// CHECK-LABEL:   func @maskedload_regression_1(
//  CHECK-SAME:       %[[BASE:.*]]: memref<?xf32>,
//  CHECK-SAME:       %[[PASS_THRU:.*]]: vector<16xf32>
func.func @maskedload_regression_1(%base: memref<?xf32>, %pass_thru: vector<16xf32>) {
  ...
  //       CHECK:   %[[LOAD:.*]] = vector.load %[[BASE]][%[[C]]] 
  %ld = vector.maskedload %base[%c0], %mask, %pass_thru
  ...
}
```
Step 2 - Improve Test Naming
```mlir
// CHECK-LABEL:   func @maskedload_to_load_dynamic_i32_all_true(
//  CHECK-SAME:       %[[BASE:.*]]: memref<?xf32>,
//  CHECK-SAME:       %[[PASS_THRU:.*]]: vector<16xf32>
func.func @maskedload_to_load_dynamic_i32_all_true(%base: memref<?xf32>, %pass_thru: vector<16xf32>) {
  ...
}
```
The doc section could open with IR blob with the original three test cases, show these reduced steps, and close with all three test cases shown again with all these steps applied.

As usual, different tradeoffs - just treat it as yet another possibility.
And of course, it is biased toward my style of reading where I prefer to do quick visual diff than read text 😉

tested, providing high-level reasoning, and consolidating shared comments. For
instance:

```mlir
///----------------------------------------------------------------------------------------
/// [Pattern: TransferWritePermutationLowering]
///
/// IN: vector.transfer_write (_transposed_ minor identity permutation map)
/// OUT: vector.transpose + vector.transfer_write (minor identity permutation map)
///
/// Note, `permutation_map` from the input Op is replaced with the newly
/// inserted vector.traspose Op.
///----------------------------------------------------------------------------------------

// CHECK-LABEL:   func.func @xfer_write_minor_identity_transposed
//  CHECK-SAME:      %[[VEC:.*]]: vector<4x8xi16>,
//  CHECK-SAME:      %[[MEM:.*]]: memref<2x2x8x4xi16>
//  CHECK-SAME:      %[[IDX:.*]]: index)
//       CHECK:      %[[TR:.*]] = vector.transpose %[[VEC]], [1, 0] : vector<4x8xi16> to vector<8x4xi16>
// (...)
```

This documents:
* The transformation pattern being tested.
* The key logic behind the transformation.
* The expected change in output.

#### Documenting the "What"
While we generally document why, documenting what is valid and encouraged in
cases where:

> * The he test generates long or complex output

Reviewer 2: ```suggestion
* The test generates long or complex output
```

// CHECK-SAME:       indexing_maps = [#[[INPUT_MAP]], #[[FILTER_MAP]], #[[OUTPUT_MAP]]],
// CHECK-SAME:       iterator_types = ["parallel", "parallel", "parallel", "reduction"]
// CHECK-SAME:       kind = #vector.kind<add>
// CHECK-SAME:     %[[V_INPUT_1]], %[[V_FILTER]], %[[V_OUTPUT_1]]
// CHECK-SAME:     : vector<4x1x3xf32>, vector<3x8xf32> into vector<4x1x8xf32>

/// w == 0, kw == 0
//      CHECK:   %[[RES_0:.+]] = vector.insert_strided_slice %[[CONTRACT_0]], %[[V_OUTPUT_R]]
// CHECK-SAME:     {offsets = [0, 0, 0], strides = [1, 1, 1]} : vector<4x1x8xf32> into vector<4x2x8xf32>
/// w == 1, kw == 0
//      CHECK:   %[[RES_1:.+]] = vector.insert_strided_slice %[[CONTRACT_1]], %[[RES_0]]
// CHECK-SAME:     {offsets = [0, 1, 0], strides = [1, 1, 1]} : vector<4x1x8xf32> into vector<4x2x8xf32>

/// Write the result back in one shot.
//      CHECK:   vector.transfer_write %[[RES_1]], %[[OUTPUT]][%[[C0]], %[[C0]], %[[C0]]]

```

Note that while the comments document what is happening (e.g., "Write the
result back in one shot"), some variables — like `w` and `kw` — remain
"enigmatic" and are not explicitly explained. This might leave us

> second-guessing their meaning at first. However, that’s intentional—their

Reviewer 2: nit: white spaces around dashes

Author: I assume that you were referring to `intentional-their`?

`second-guessing` appears to be correct: https://dictionary.cambridge.org/dictionary/english/second-guess

Btw, English is my 2nd language -  I am always happy to be corrected :)

Reviewer 2: Oh yes, I mean the former. There was another one in a different paragraph that you already fixed with other changes.

The update in Step 2 made us realize that we were missing an important edge
case:

* A mask that is neither "all true" nor "all false".

Unlike the existing cases, this mask must be preserved. In this scenario,
`vector.load` is not the right abstraction. Thus, no lowering should occur:

```mlir
// CHECK-LABEL:   func @negative_maskedload_to_load_static_i32_mixed(
// CHECK-SAME:        %[[BASE:.*]]: memref<16xf32>,
// CHECK-SAME:        %[[PASS_THRU:.*]]: vector<16xf32>) -> vector<16xf32> {
//      CHECK:    vector.maskedload
func.func @negative_maskedload_to_load_static_i32_mixed(%base: memref<16xf32>, %pass_thru: vector<16xf32>) -> vector<16xf32> {
  %c0 = arith.constant 0 : index
  %mask = vector.constant_mask [4] : vector<16xi1>

  %ld = vector.maskedload %base[%c0], %mask, %pass_thru
    : memref<16xf32>, vector<16xi1>, vector<16xf32> into vector<16xf32>

  return %ld : vector<16xf32>
}
```

The `negative_` prefix indicates that this test should fail to lower, as the
pattern should not match.

To summarize, here’s the naming convention used:

* `@{negative_}?maskedload_to_load_{static|dynamic}_{i32|i8}_{all_true|all_false|mixed}`.

> #### What if there is no pre-existing style to follow?

Reviewer: I like this section, it aligns pretty close with what I had in mind in one of the comment threads. Thanks!

If you are adding a new test file, you can use other test files in the same
directory as inspiration.

If the test file you are modifying lacks a clear style and instead has mixed,
inconsistent styles, try to identify the dominant one and follow it. Even

> better, consider refactoring the file to adopt a single, consistent style —

Reviewer: I wonder if it would be worth spelling out that such refactoring should generally go to a separate PR, or if that's obvious.

Author: This is something I'd leave under-specified.

I agree with you that such changes should be submitted as a separate PR. At least in general. However, there are downsides:
* more churn + more PR traffic.

Also, sometimes the noise level is fairly low and sending a dedicated PR might be too much. So, I'd trust our judgement on case-by-case basis. Is that OK? We can always refine later.

this helps improve our overall testing quality. Alternatively, create a GitHub

> issue and leave a TODO in the test file taht you are modyfying.

Reviewer: ```suggestion
issue and leave a TODO linking to it in the test file that you are modifying.
```

This is also encouraged when the existing style leaves room for improvement.

When creating a new naming convention, keep these points in mind:

* **Write Orthogonal Tests**
Struggling with good naming may indicate the absence of a clear pattern in the
set of tests being written. A good rule of thumb is to avoid testing the same
thing repeatedly. Before writing tests, define clear categories to cover (e.g.,
number of loops, data types). This often leads to a natural naming scheme—for
example: `@loop_depth_2_i32`.

* **What vs Why**
Test names should reflect _what_ is being tested, not _why_.

Encoding _why_ in test names can lead to unnecessarily long and complex names.

> Instead, add inline comments where needed—but avoid over-explaining.

Reviewer: ```suggestion
Instead, add inline comments where needed.
```
nit: I think this already implies our preference for conciseness.

#### Don't forget the common sense

As a reviewer or contributor, always apply common sense when naming functions
and variables. Encoding too much information in names can negatively impact
readability and maintainability.

Trust your judgment. When in doubt, consult your future self: "Will this still
make sense to me six months from now?"

#### Final Points - Key Principles

The above approach is just an example. It may not fit your use case perfectly,
so feel free to adapt it as needed. For example, for "folding" tests, it may
make more sense to use "no" as prefix for negative tests (e.g.
`@no_fold_<case>_<subcase>`). Key principles to follow:

* Make tests self-documenting.
* Follow existing conventions.

By applying these best practices, we leverage available tools (e.g., test
function names) to make tests easier to discover and maintain.

> ### Test Docmentation Best Practices

Reviewer: ```suggestion
### Test Documentation Best Practices
```

/// Expect the in_bounds attribute to be preserved. However, since we don't
//// print it when all flags are "false", it should not appear in the output.
/// CHECK-NOT:       in_bounds

// CHECK:           vector.transfer_write

/// The permutation map was replaced with vector.transpose
// CHECK-NOT:       permutation_map

// CHECK-SAME:      %[[TR]], %[[MEM]][%[[IDX]], %[[IDX]], %[[IDX]], %[[IDX]]] : vector<8x4xi16>, memref<2x2x?x?xi16>
func.func @xfer_write_minor_identity_transposed_out_of_bounds(
    %vec: vector<4x8xi16>,
    %mem: memref<2x2x?x?xi16>,
    %idx: index) {

  vector.transfer_write %vec, %mem[%idx, %idx, %idx, %idx] {
    in_bounds = [false, false],
    permutation_map = affine_map<(d0, d1, d2, d3) -> (d3, d2)>
  } : vector<4x8xi16>, memref<2x2x?x?xi16>

  return
}
```

Here, we document two non-obvious behaviors:

* _Why_ is `permutation_map` missing from the output?
* _Why_ is the `in_bounds` attribute missing in one of the outputs?

#### How to Identify What Needs Documentation?
A good rule of thumb is to think of yourself six months from now and ask:
"What might be difficult to decipher without comments?"

> If you expect something to be tricky for future-you, it’s likely to be tricky

Reviewer: ```suggestion
If you expect something to be tricky for the future-you, it’s likely to be tricky
```
